cpp/src/arrow/csv/csv-to-arrow.cc

```diff
@@ -0,0 +1,162 @@
+namespace csv {
+
+static Status InitParseOptions(ParseOptions& options) {
+  options.delimiter = FLAGS_delimiter.at(0);
+  options.quoting = FLAGS_quoting;
+  options.quote_char = FLAGS_quote_char.at(0);
+  options.double_quote = FLAGS_double_quote;
+  options.escaping = FLAGS_escaping;
+  options.escape_char = FLAGS_escape_char.at(0);
+  options.newlines_in_values = FLAGS_newlines_in_values;
+  options.ignore_empty_lines = FLAGS_ignore_empty_lines;
+  options.header_rows = FLAGS_header_rows;
+  return Status::OK();
+}
+
+static Status InitConvertOptions(ConvertOptions& options) {
+  options.check_utf8 = FLAGS_check_utf8;
+  return Status::OK();
+}
+
+static Status InitReadOptions(ReadOptions& options) {
+  options.use_threads = FLAGS_use_threads;
+  options.block_size = FLAGS_block_size;
+  return Status::OK();
+}
+
+static Status Run(int argc, char** argv) {
+  // Instantiate and handle options
+  gflags::SetUsageMessage("csv-to-arrow [FLAGS] input.csv  > output.arrow");
+  gflags::ParseCommandLineFlags(&argc, &argv, true);
+  if (argc != 2) {
+    std::cerr << "Usage: " << gflags::ProgramUsage() << std::endl;
+    return Status::Invalid("Input filename missing");
+  }
+  auto parse_options = ParseOptions::Defaults();
+  ARROW_RETURN_NOT_OK(InitParseOptions(parse_options));
+  auto read_options = ReadOptions::Defaults();
+  ARROW_RETURN_NOT_OK(InitReadOptions(read_options));
+  auto convert_options = ConvertOptions::Defaults();
+  ARROW_RETURN_NOT_OK(InitConvertOptions(convert_options));
+
+  // Instantiate reading
+  std::string input_filename(argv[1]);
+  std::shared_ptr<io::ReadableFile> input_file;
+  ARROW_RETURN_NOT_OK(io::ReadableFile::Open(input_filename, &input_file));
+  auto pool = default_memory_pool();
+  std::shared_ptr<TableReader> reader;
+  ARROW_RETURN_NOT_OK(TableReader::Make(pool, input_file, read_options, parse_options,
+                                        convert_options, &reader));
+
+  // Instantiate writing, read from input and write to output
+  io::StdoutStream stdout;
+  std::shared_ptr<ipc::RecordBatchWriter> writer;
+  bool schema_written = false;
+  std::shared_ptr<arrow::Table> table;
+  while (!reader->Read(&table).IsInvalid()) {
+    if (!schema_written) {
+      RETURN_NOT_OK(ipc::RecordBatchFileWriter::Open(&stdout, table->schema(), &writer));
+      schema_written = true;
+    }
+    std::shared_ptr<RecordBatch> batch_out;
+    TableBatchReader table_reader(*table);
+    while (table_reader.ReadNext(&batch_out).ok()) {
+      if (batch_out == nullptr) {
+        break;
+      }
+      ARROW_RETURN_NOT_OK(writer->WriteRecordBatch(*batch_out));
+    }
+  }
+
+  // Cleanup
+  ARROW_RETURN_NOT_OK(writer->Close());
+  gflags::ShutDownCommandLineFlags();
+  return Status::OK();
+}
+
+}  // namespace csv
+}  // namespace arrow
+
+int main(int argc, char** argv) {
+  arrow::Status status = arrow::csv::Run(argc, argv);
+  if (!status.ok()) {
+    std::cerr << "ERROR! " << status.ToString() << std::endl;
```

Comment:
You also want to return a non-zero error code here.

Reply:
Yep!